regionGrowing.py

```diff
@@ -0,0 +1,192 @@
+        return 0
+    else:
+        return np.sqrt(var)
+
+def pointsInRadius(sPoint, radius, dims=2):
+    points = []
+    if radius == 0:
+        points = [sPoint]
+    else:
+        # Dodanie pionowych
+        for row in [sPoint[0] - radius, sPoint[0] + radius]:
+            [points.append([row, col]) for col in range(sPoint[1] - radius, sPoint[1] + radius + 1)]
+        
+        # Dodanie poziomych
+        for col in [sPoint[1] - radius, sPoint[1] + radius]:
+            [points.append([row, col]) for row in range(sPoint[0] - radius + 1, sPoint[0] + radius)]
+
+    return points
+
+def belongsToArr(p, arr):
+    """
+    Sprawdza czy punkt lub zbiór punktów należy(należą) do macierzy.
+    Jeśli p: punkt (2D lub 3D) to:
+        True, jeśli punkt należy do macierzy
+        False wpp
+    Jeśli p: zbiór punktów (2D lub 3D) to:
+        True, jeśli PRZYNAJMNIEJ JEDEN punkt należy do macierzy
+        False, jeśli ŻADEN punkt nie należy do macierzy 
+    """
+    p = np.array(p)
+    if len(p.shape) == 1:
+        # punkt
+        if len(p) == 2:     # 2D
+            # Dorobić wyjątki gdy wymiary się nie zgadzają
+            logic = p[0] >= 0 and p[0] < arr.shape[0] and p[1] >= 0 and p[1] < arr.shape[1]
+            if logic:
+                return True
+            else:
+                return False
+        
+        else:       # 3D
+            logic = p[0] >= 0 and p[0] < arr.shape[0] and p[1] >= 0 and p[1] < arr.shape[1] and p[2] >= 0 and p[2] < arr.shape[2]
+            if logic:
+                return True
+            else:
+                return False
+        
+    else:
+        # zbiór punktów. Sprawdzamy, czy choć jeden punkt/woksel jest na macierzy
+        if len(p[0]) == 2:  # 2D
+            for point in p:
+                logic = point[0] >= 0 and point[0] < arr.shape[0] and point[1] >= 0 and point[1] < arr.shape[1]
+                if logic:
+                    return True
+            return False
+        
+        else:       # 3D
+            for point in p:
+                logic = point[0] >= 0 and point[0] < arr.shape[0] and point[1] >= 0 and point[1] < arr.shape[1] and point[2] >= 0 and point[2] < arr.shape[2]
+                if logic:
+                    return True
+           
+            return False
+
+
+# Dane wejściowe
+# arr =np.random.randint(100, size=(100,100))
+# thresh = 50
+# startPoint = (0,0)
+# arr[startPoint] = 51
+c = 1
+
+# Już na serio
+path = '/home/krzysztof/Dokumenty/Praca_inż/j-pet-roi-detector/fantomNowy.pckl'
+vol = VolumeData(path)
+
+arr = vol.getSlice(75)
+thresh = threshold_yen(arr)
+bw = closing(arr > thresh, square(3))
+label_image = label(bw)
+
+centroidy = []
+for region in regionprops(label_image):
+    # take regions with large enough areas
+    if region.area >= 10:
+        centroidy.append(region.centroid)
+
+centroidy = np.array(centroidy)
+startPoints = np.around(centroidy).astype(int)
+startPoint= startPoints[2]
```

Comment:
Do tego miejsca wygląda OK

Czemu `startPoints[2]` ??

Co do samego algorytmu to chyba nie ma sensu ograniczać się do wyszukiwania w coraz rosnącym promieniu:
- Będzie ciężko stwierdzić które punkty powinny należeć do regionu a które nie (które są połączone)
- Dla regionów które nie są symetryczne będziemy przemiatali sporo zbednych punktów

Najlepiej chyba użyć jakiejś implementacji `dilate` np: https://scikit-image.org/docs/dev/api/skimage.morphology.html#skimage.morphology.dilation

Dostaniesz region powiększony np o jeden piksel w każdym kierunku. Wtedy moższ nałożyć na niego progowanie. Po progowaniu wyznaczyć średnią i odchylenie używając maski. Następnie powtórzyć algorytm.

Reply:
Chyba nie za bardzo rozumiem idei. W tym momencie tworzony obszar trzymam jako listę. Idąc po spirali dodaję kolejne punkty do listy w zależności od tego czy spełniają równanie czy nie. Na jakim obiekcie miałbym wykonać dylację ? Czy chodzi o to, żeby stworzyć pustą macierz o wymiarach badanego wolumenu i na niej wykonywać dylację, a punkty dodane przez dylację następnie przepuszczać przez równanie ? To mi się wydaje o wiele bardziej nieekonomiczne pamięciowo, ponieważ mielibyśmy pustą w 99% macierz 160x160x120. 
Tymczasem, gdy używam punktów z coraz większego promienia, to mam tam warunke stopu: gdy żaden z punktów w promieniu R nie zostanie dodany do obszaru, algorytm przerywa działanie. I w ten sposób maksymalny promień jaki udało mi się osiągnąć dla fantoma to 4.